content/exchange/artifacts/Windows.Powershell.Timeline.yaml

```diff
@@ -0,0 +1,102 @@
+name: Windows.Powershell.Timeline
+author: Mohammed Hasan, 0xHasanm
+description: |
+   This artifact extracts commands executed from each user's ConsoleHost_History.txt and correlates the timestamps of Data_Added events in the USNJRNL with those commands to construct a timeline of PowerShell command execution.
+
+# Can be CLIENT, CLIENT_EVENT, SERVER, SERVER_EVENT or NOTEBOOK
+type: CLIENT
+
+parameters:
+   - name: user
+     default: .
+
+references:
+  - https://www.linkedin.com/posts/0xhasanm_sundfirday-dfir-powershell-activity-7395908066409484288-eNSo
+
+sources:
+  - precondition:
+      SELECT OS From info() where OS = 'windows'
+
+    query: |
+      LET USN_EVENTS <= SELECT
+          Timestamp,
+          split(string=OSPath, sep_string="\\")[-8] AS username,
+          Reason
+        FROM Artifact.Windows.Forensics.Usn(FileNameRegex="ConsoleHost_History.txt")
+        WHERE username =~ user
+         and len(list=Reason) = 2
+              and (Reason[0] = "DATA_EXTEND" or Reason[1] = "DATA_EXTEND")
+        ORDER BY Timestamp DESC
+      
+      LET PSReadline <= SELECT LineNum,
+                               Line,
+                               Username
+        FROM Artifact.Windows.System.Powershell.PSReadline()
+        WHERE Username =~ user
+        ORDER BY LineNum DESC
+      
+      LET USN_EVENTS_Indexed <= SELECT count() AS EventID,
+                                       *
+        FROM USN_EVENTS
+      
+      LET PSReadline_Indexed <= SELECT count() AS CommandID,
+                                       *
+        FROM PSReadline
+      
+      LET results_with_timestamp <= SELECT *
```

Comment:
This will be very slow. It looks like you want to do a join. See this

https://docs.velociraptor.app/docs/vql/join/